### PR TITLE
provider/aws: Add timestamps to final snapshot

### PR DESCRIPTION
**Pull request by netjunki**
tl;dr: add a `timestamp_final_snapshot` parameter to `aws_db_instance` and `aws_redshift_cluster`. When deleting the instance this generates a snapshot named `<final_snapshot_identifier>-<timestamp>`

Example: generates a snapshot called mydb-20170210-083019

    resource "aws_db_instance" "default" {
      allocated_storage    = 10
      engine               = "mysql"
      instance_class       = "db.t1.micro"
      name                 = "mydb"
      username             = "foofoofoo"
      password             = "barbarbar"
      db_subnet_group_name = "mydb_subnet_group"
      parameter_group_name = "default.mysql5.6"
      final_snapshot_identifier = "mydb"
      timestamp_final_snapshot = "20060102-150405"
    }

My team have been encountering the following error intermittently the past couple of weeks:
    4 error(s) occurred:
    
    * aws_db_instance.default: DBSnapshotAlreadyExists: Cannot create the snapshot because a snapshot with the identifier finalsnapshotcifull-service-mysql already exists.
        status code: 400, request id: 456a579a-ee6a-11e6-a5b8-cdbcbd0418d6

We've also been encountering these with redshift clusters now that I've started to use them:
    
    * aws_redshift_cluster.redshift-cluster: [ERROR] Error deleting Redshift Cluster (bentest): ClusterSnapshotAlreadyExists: Cannot create the snapshot because a snapshot with the identifier bentest-final-snapshot already exists.

To help mitigate the issue we had been appending a `md5(timestamp())` to the final_snapshot_identifier parameter of aws_db_instance, but this was causing some less than optimal behavior when running terraform apply against the same environment repeatedly, not to mention the fact that the snapshot name wasn't particularly readable.

To mitigate the issue I decided it would be useful to have a real timestamp appended to the snapshot identifier. It uses a [golang time Format](https://golang.org/pkg/time/#Time.Format) string to determine the formatting.

### CODE REVIEW

**Comment by avatar4d** [comment]
+1

**Comment by panmanphil** [comment]
+1

**Comment by shoelessrob** [comment]
+1

**Comment by stack72** [comment]
Hi @netjunki 

So i am in agreement of this feature - I do think we are going to need an acceptance test to show that it works as expected

Apart from that, the code looks good

Paul

**Comment by netjunki** [comment]
@stack72 Sure. What do I need to do to build out an acceptance test for this?

**Comment by stack72** [comment]
@netjunki there is a short guide on writing them here https://github.com/hashicorp/terraform/blob/master/.github/CONTRIBUTING.md#writing-acceptance-tests

**Comment by netjunki** [comment]
@stack72 busy week so didn't get a chance to work on this until today. I just pushed an acceptance test for the aws_db_instance and wanted to get a review before I apply the same approach to redshift clusters.

The timestamp is date based... so theoretically this is going to have some sort of failure if the test starts on one day and finished during the next. Is there some sort of date mocking in the terraform acceptance testing framework?

**Comment by netjunki** [comment]
@stack72 If you could review this that'd be awesome. I'd like to try and wrap up this feature this week.

**Comment by stack72** [comment]
@netjunki do you think it would be better to make the schema for this a bool? Then we can make it false by default and then get users to toggle it on when they want a timestamp appended?

**Comment by netjunki** [comment]
@stack72 Well we could do that, but the goal I was shooting for was to allow the timestamp to be configurable. Not everyone will want a YYYYMMDD-HHMMSS timestamp... or at least I didn't feel like imposing my choice on others. :-) We could split it into two options. One for whether it has a timestamp (with a default of YYYYMMDD-HHMMSS) as a boolean and a second one for the format. But two options seemed a bit much for this.

**Comment by netjunki** [comment]
Yeah that's cool with me. I can probably get that hammered out in the next
day or two.

So we'd have two options:
  timestamp_final_snapshot - boolean determines whether to timestamp or not
 timestamp_final_snapshot_format - format of the timestamp with a default
of "20060102-150405" unless specified.

Also, if you could review the approach I took with the acceptance test I
can get that updated as well and create the tests for the redshift cluster
as well.

Please adjust the names above/default timestamp format if you have
different ideas.

**Comment by grubernaut** [comment]
@netjunki just checking, but what would stop you from using the `timestamp()` interpolation function in the snapshot identifier? 
```
final_snapshot_identifier = "mydb-${timestamp()}"
```

**Comment by netjunki** [comment]
@grubernaut it regenerates it with the current time when you run terraform... or at least I'm pretty sure that's what it was doing. So every run you get a new value which means terraform thinks there are changes even though I'd like it to report none. Also its format isn't valid for the snapshot names in aws. We work around that now by using ${md5(timestam())} but these aren't sortable. Also... then the timestamp would be the time we ran terraform... and not when the final snapshot was generated. This is how this who PR started. :-)

**Comment by netjunki** [comment]
@grubernaut @stack72 How does the aws_db_instance half of this look now? If it looks okay I'll make the redshift cluster match, add an acceptance test and I think this should be ready to go. :-)

**Comment by netjunki** [comment]
Went ahead and just did the redshift cluster updates since I was waiting around for a day job task to complete.

**Comment by netjunki** [comment]
@stack72 @grubernaut any time to get a review on this? (probably got your hands full with 0.9.2...) :-)

**Comment by netjunki** [comment]
@stack72 @grubernaut anything else required for this?

**Comment by netjunki** [comment]
@stack72 @grubernaut bump (haven't had cycles to keep pushing on this for a couple of weeks)

**Comment by apparentlymart** [comment]
Hello @netjunki, and thanks for working on this!

As part of the the Terraform 0.10 release earlier this year, all of the Terraform providers were moved to their own repositories in [the terraform-providers GitHub organization](https://github.com/terraform-providers), and removed from the Terraform Core repository.

Unfortunately due to the fact that new issues and pull requests are being opened constantly, it was not possible for the various provider maintainers to merge all outstanding pull requests before this split, and there is no automatic way to migrate a pull request to a new repository.

As a result, this pull request can sadly no longer be applied as-is, and so I'm going to close it.

If you or someone else has the time and motivation to apply same changes to [the `aws` provider repository](https://github.com/terraform-providers/terraform-provider-aws) and open a new PR there, the maintainers of that provider should be able to review and merge it.

Thanks again for working on this, and sorry it was not able to be merged before the provider repository changes.
